mythril/analysis/modules/dos.py

```diff
@@ -0,0 +1,111 @@
+"""This module contains the detection code for insecure delegate call usage."""
+
+import logging
+from typing import Dict, cast, List
+
+from mythril.analysis.swc_data import DOS_WITH_BLOCK_GAS_LIMIT
+from mythril.analysis.report import Issue
+from mythril.analysis.modules.base import DetectionModule
+from mythril.laser.ethereum.state.global_state import GlobalState
+from mythril.laser.ethereum.state.annotation import StateAnnotation
+from mythril.laser.ethereum import util
+
+log = logging.getLogger(__name__)
+
+
+class LoopAnnotation(StateAnnotation):
+    def __init__(self, jump_address: int) -> None:
+        self.jump_address = jump_address
+
+
+class DOS(DetectionModule):
+    """This module detects denial of service by exhausting gas."""
+
+    def __init__(self) -> None:
+        """"""
+        super().__init__(
+            name="DOS",
+            swc_id=DOS_WITH_BLOCK_GAS_LIMIT,
+            description="Check for DOS",
+            entrypoint="callback",
+            pre_hooks=["JUMPI", "CALL", "SSTORE"],
+        )
+
+        """Keeps track of how often jump destinations are reached."""
+        self._jumpdest_count = {}  # type: Dict[object, dict]
+
+    def execute(self, state: GlobalState) -> None:
+        """
+
+        :param state:
+        :return:
+        """
+        log.debug("Executing module: DOS")
+
+        self._issues.extend(self._analyze_states(state))
+
+    def _analyze_states(self, state: GlobalState) -> List[Issue]:
+        """
+        :param state: the current state
+        :return: returns the issues for that corresponding state
+        """
+
+        opcode = state.get_current_instruction()["opcode"]
+
+        if opcode == "JUMPI":
+
+            target = util.get_concrete_int(state.mstate.stack[-1])
+
+            try:
+                transaction = state.current_transaction
+                self._jumpdest_count[transaction]
+
+                try:
+                    if self._jumpdest_count[transaction][target] > 3:
+                        state.annotate(
```

Comment:
I think we might be able to reduce the number of annotate()'s here, by checking if it's 4 then annotate and ++, >4 then ignore, might save us some memory.